payu/fsops.py

```diff
@@ -229,3 +234,93 @@ def list_archive_dirs(archive_path: Union[Path, str],
 
     dirs.sort(key=lambda d: int(d.lstrip(dir_type)))
     return dirs
+
+
+def run_script_command(script_cmd: str, control_path: Path) -> None:
+    """Run a user defined script or command.
+
+    Parameters
+    ----------
+    script_cmd : string
+        String of user-script command defined in configuration file
+    control_path : Path
+        The control directory of the experiment
+
+    Raises
+    ------
+    RuntimeError
+        If there's was an error running the user-script
+    """
+    try:
+        _run_script(script_cmd, control_path)
+    except Exception as e:
+        error_msg = f"User defined script/command failed to run: {script_cmd}"
+        raise RuntimeError(error_msg) from e
+
+
+def needs_subprocess_shell(command: str) -> bool:
+    """Check if command contains shell specific values. For example, file
+    redirections, pipes or logical operators.
+
+    Parameters
+    ----------
+    command: string
+        String of command to run in subprocess call
+
+    Returns
+    -------
+    bool
+        Returns True if command requires a subprocess shell, False otherwise
+    """
+    shell_values = ['>', '<', '|', '&&', '$', '`']
+    for value in shell_values:
+        if value in command:
+            return True
+    return False
+
+
+def _run_script(script_cmd: str, control_path: Path) -> None:
+    """Helper recursive function to attempt running a script command.
+
+    Parameters
+    ----------
+    script_cmd : string
+        The script command to attempt to run in subprocess call
+    control_path: Path
+        The control directory to use for resolving relative filepaths, if file
+        is not found
+    """
+    # First try to interpret the argument as a full command:
+    try:
+        if needs_subprocess_shell(script_cmd):
+            subprocess.check_call(script_cmd, shell=True)
+        else:
+            subprocess.check_call(shlex.split(script_cmd))
+    except EnvironmentError as e:
```

Comment:
Adding random late night thoughts - this will likely require a revisit:
* Not having seen `EnvironmentError` before, I discovered it's [an alias of OSError](https://docs.python.org/3/library/exceptions.html#EnvironmentError)
* `errno.ENOENT` [maps to FileNotFoundError](https://docs.python.org/3/library/errno.html#errno.ENOENT)
* `errno.EACCES` [maps to PermissionError](https://docs.python.org/3/library/errno.html#errno.EACCES)

Would this func look better if reworked to a structure like:
```Python
try:
    # subprocess stuff
except OSError:
    # handler
except FileNotFoundError:
    # handler
except PermissionError:
    # handler

# then else or finally block if needed?
```
Another thought: would it be cleaner to detect the input type outside the exception handler block, then call the relevant processing func?

Reply:
Thanks, I've updated the function with the using `FileNotFoundError` and `PermissionError` and it does look cleaner!

Reply:
> Another thought: would it be cleaner to detect the input type outside the exception handler block, then call the relevant processing func? 

Do you mean like instead of the `FileNotFoundError`, check first if the cmd is a file in the control path?

Reply:
> > Another thought: would it be cleaner to detect the input type outside the exception handler block, then call the relevant processing func?
> 
> Do you mean like instead of the `FileNotFoundError`, check first if the cmd is a file in the control path?

Yep, I imagined something like
```Python
def detect_script_type(script_cmd):
    if is_a_script_path(script_cmd):  # runs some detection mechanism
        return something  # return a function that runs scripts???
    else:
        # check for user commands etc.
        return something_else

def workflow(script_cmd):
    # ... processing ...
    func = detect_script_type(script_cmd)
    func(script_cmd)
    # ... processing ...    
```
This may not gain much beyond simply extracting some logic from the new exception handler blocks. Do you see any other reason(s) to modify the exception block approach?

Reply:
Yeah I briefly had a go at that, but then ended writing a bit more code than the current recursive approach.